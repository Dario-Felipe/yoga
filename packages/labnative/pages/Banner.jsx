import React from 'react';

import styled from 'styled-components';

import { Banner } from '@gympass/yoga';

import { DocTitle } from '../components';

const ScrollView = styled.ScrollView`
  margin-bottom: 50px;
  padding: 10px;
  width: 100%;
`;

const BannerWrapper = styled.View`
  width: 100%;
  padding: 20px;
  margin-bottom: 10px;
  align-items: center;
  justify-content: space-between;
  height: 300px;
  flex-direction: column;
`;

const BannerPage = () => (
  <ScrollView>
    <BannerWrapper>
      <DocTitle>Banner</DocTitle>
      <Banner
        variant="success"
        message="Success Banner"
        button={{ label: 'action', action: () => {} }}
      />
      <Banner
        variant="informative"
        message="Informative banner"
        button={{ label: 'action', action: () => {} }}
      />
      <Banner
        variant="attention"
        message="Attention banner"
        button={{ label: 'action', action: () => {} }}
      />
    </BannerWrapper>

    <BannerWrapper>
      <DocTitle>Banner without button</DocTitle>
      <Banner variant="success" message="Success banner" />
      <Banner variant="informative" message="Informative banner" />
      <Banner variant="attention" message="Attention banner" />
    </BannerWrapper>
  </ScrollView>
);

export default BannerPage;
