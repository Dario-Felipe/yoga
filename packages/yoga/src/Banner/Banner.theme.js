const Banner = ({ colors, spacing, radii }) => ({
  defaultBackgroundColor: colors.feedback.informative.light,
  padding: {
    top: spacing.small,
    right: spacing.small,
    bottom: spacing.small,
    left: spacing.small,
  },
  border: {
    radius: radii.small,
  },
});

export default Banner;
