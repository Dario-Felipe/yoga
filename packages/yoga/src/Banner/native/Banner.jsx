import React from 'react';

import styled from 'styled-components';
import { margins } from '@gympass/yoga-system';
import { func, oneOf, shape, string } from 'prop-types';

import Text from '../../Text';
import Button from '../../Button';

const StyledBanner = styled.View`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: center;
  ${({
    variant,
    theme: {
      yoga: {
        components: { banner },
        colors: {
          feedback: {
            [variant]: backgroundColor = {
              light: banner.defaultBackgroundColor,
            },
          },
        },
      },
    },
  }) => `
    background-color: ${backgroundColor.light};
    padding: ${banner.padding.top}px
      ${banner.padding.right}px
      ${banner.padding.bottom}px
      ${banner.padding.left}px;
    border-radius: ${banner.border.radius}px;
  `}

  ${margins}
`;

const Banner = ({ variant, message, button }) => (
  <StyledBanner variant={variant}>
    <Text.Small flex={1}>{message}</Text.Small>
    {!!button && (
      <Button.Link marginLeft="small" secondary onPress={button.action}>
        {button.label}
      </Button.Link>
    )}
  </StyledBanner>
);

Banner.propTypes = {
  variant: oneOf(['success', 'informative', 'attention']),
  message: string.isRequired,
  button: shape({
    label: string.isRequired,
    action: func.isRequired,
  }),
};

Banner.defaultProps = {
  variant: 'informative',
  button: null,
};

export default Banner;
