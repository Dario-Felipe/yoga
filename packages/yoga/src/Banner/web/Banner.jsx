import React from 'react';

import styled from 'styled-components';
import { margins, paddings } from '@gympass/yoga-system';
import { checkPropTypes, func, oneOf, string } from 'prop-types';

import Text from '../../Text';
import Button from '../../Button';
import Box from '../../Box';

const StyledBanner = styled(Box)`
  ${({
    variant,
    theme: {
      yoga: {
        components: { banner },
        spacing,
        colors: {
          feedback: {
            [variant]: backgroundColor = {
              light: banner.defaultBackgroundColor,
            },
          },
        },
      },
    },
  }) => `
    background-color: ${backgroundColor.light};
    padding: ${spacing.xsmall}px
      ${spacing.small}px;
    border-radius: ${banner.border.radius}px;
  `}

  ${margins}

  ${paddings}
`;

/** A banner is a component that displays a prominent message. It can have a related actions button on it or not. */
const Banner = ({ variant, message, actionLabel, onAction, ...props }) => (
  <StyledBanner
    display="flex"
    flex={1}
    alignItems="center"
    of="hidden"
    variant={variant}
    {...props}
  >
    <Text.Small flex={1} marginVertical="xxxsmall">
      {message}
    </Text.Small>
    {!!onAction && actionLabel && (
      <Box
        as={Button.Text}
        marginLeft="xxsmall"
        small
        secondary
        onClick={onAction}
      >
        {actionLabel}
      </Box>
    )}
  </StyledBanner>
);

Banner.propTypes = {
  /** style the banner following the theme (success, informative, attention) */
  variant: oneOf(['success', 'informative', 'attention']),
  message: string.isRequired,
  /** Function for the custom action. The `actionLabel` becomes required when passing this function. */
  onAction: (props, propName, componentName) => {
    const { actionLabel } = props;

    if (actionLabel) {
      checkPropTypes(
        { [propName]: func.isRequired },
        props,
        'prop',
        componentName,
      );
    }

    return null;
  },
  /** Label for a custom action. */
  actionLabel: (props, propName, componentName) => {
    const { onAction } = props;

    if (onAction) {
      checkPropTypes(
        { [propName]: string.isRequired },
        props,
        'prop',
        componentName,
      );
    }

    return null;
  },
};

Banner.defaultProps = {
  variant: 'informative',
  actionLabel: null,
  onAction: null,
};

export default Banner;
