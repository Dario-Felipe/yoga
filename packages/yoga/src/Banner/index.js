import Banner from './web';

export default Banner;
