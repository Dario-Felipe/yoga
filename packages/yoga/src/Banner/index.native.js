import Banner from './native';

export default Banner;
