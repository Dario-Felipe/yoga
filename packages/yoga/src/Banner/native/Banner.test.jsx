import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import { ThemeProvider } from '../..';
import Banner from './Banner';

describe('<Banner />', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <ThemeProvider>
        <Banner message="Feedback message." />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with custom variant', () => {
    const { container } = render(
      <ThemeProvider>
        <Banner variant="success" message="Success banner without button" />
        <Banner
          variant="informative"
          message="Informative banner without button"
        />
        <Banner variant="attention" message="Attention banner without button" />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with button', () => {
    const { container } = render(
      <ThemeProvider>
        <Banner
          message="Banner with button"
          button={{ label: 'Action', action: () => {} }}
        />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should call button action function when the banner button is pressed', () => {
    const onButtonPressMock = jest.fn();
    const { getByText } = render(
      <ThemeProvider>
        <Banner
          message="Banner with button"
          button={{ label: 'Action', action: onButtonPressMock }}
        />
      </ThemeProvider>,
    );

    fireEvent.press(getByText('Action'));

    expect(onButtonPressMock).toHaveBeenCalled();
  });
});
