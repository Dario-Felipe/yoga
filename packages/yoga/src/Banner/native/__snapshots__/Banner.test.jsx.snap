// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should match snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0DFFF",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
      ]
    }
    variant="informative"
  >
    <Text
      flex={1}
      style={
        Array [
          Object {
            "color": "#231B22",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Rubik",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 20,
          },
        ]
      }
    >
      Feedback message.
    </Text>
  </View>
</View>
`;

exports[`<Banner /> should match snapshot with button 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0DFFF",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
      ]
    }
    variant="informative"
  >
    <Text
      flex={1}
      style={
        Array [
          Object {
            "color": "#231B22",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Rubik",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 20,
          },
        ]
      }
    >
      Banner with button
    </Text>
    <TouchableWithoutFeedback
      marginLeft="small"
      secondary={true}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <Text
          disabled={false}
          marginLeft="small"
          pressed={false}
          secondary={true}
          style={
            Array [
              Object {
                "color": "#231B22",
                "fontFamily": "Rubik",
                "fontSize": 16,
                "fontWeight": "500",
                "marginLeft": 16,
                "textAlign": "center",
              },
            ]
          }
        >
          Action
        </Text>
      </View>
    </TouchableWithoutFeedback>
  </View>
</View>
`;

exports[`<Banner /> should match snapshot with custom variant 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#C1EEDB",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
      ]
    }
    variant="success"
  >
    <Text
      flex={1}
      style={
        Array [
          Object {
            "color": "#231B22",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Rubik",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 20,
          },
        ]
      }
    >
      Success banner without button
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#E0DFFF",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
      ]
    }
    variant="informative"
  >
    <Text
      flex={1}
      style={
        Array [
          Object {
            "color": "#231B22",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Rubik",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 20,
          },
        ]
      }
    >
      Informative banner without button
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FCD6C3",
          "borderRadius": 8,
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "center",
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        },
      ]
    }
    variant="attention"
  >
    <Text
      flex={1}
      style={
        Array [
          Object {
            "color": "#231B22",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontFamily": "Rubik",
            "fontSize": 14,
            "fontWeight": "400",
            "lineHeight": 20,
          },
        ]
      }
    >
      Attention banner without button
    </Text>
  </View>
</View>
`;
